### ISSUE TITLE
is there any methods to speed up mcts?

### ISSUE THREAD

**Issue by computer-idol** [maxpumperla/deep_learning_and_the_game_of_go#68]
when I run mcts,the speed is a little slow,9*9board can only simulate 500 games in 15 seconds. In my go playing website, I have to decrease the number of  simulated games so that can receive response in limited time，but the level is also decreased.Is it possible to implement the parallel mcts?

**Comment by maxpumperla** [comment]
@computer-idol I agree, our implementation isn't optimized for speed at all, rather for educational clarity (or so we hope). If you want to help implementing a parallel version, that'd be cool! I assume that wouldn't be too hard. You could start by looking at the `multiprocessing` routines we used for processing the boards. I would love to look at this myself, but frankly don't have the time right now.

**Comment by macfergus** [comment]
Hey @computer-idol, I actually put together a C++ implementation of the `GoBoard` / `GameState` library (with Python wrapper). It's a LOT faster than the pure Python version in the book. On my laptop on a 9x9, it can simulate about 10000 games in 15 seconds. So something like 10x - 20x speedup from the book version.

I think the library fully works, but it is not fully documented and packaged yet. But if you have some patience to try building it from source, give it a try!

https://github.com/macfergus/baduk

**Comment by macfergus** [comment]
also, on the subject of parallel MCTS, it turns out to be fairly complicated, see this paper for an overview of the some of the options and tradeoffs:

http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.159.4373&rank=1

That paper was written about classical MCTS (random rollouts). With deep learning MCTS (value network), you have an additional problem where if you do two rollouts in parallel, they will give you the exact same result! So some of the methods described in the linked paper don't work at all for deep learning + MCTS, but something like the "virtual losses" method should be possible.

**Comment by computer-idol** [comment]
@macfergus wonderful!! I will try soon.

**Comment by computer-idol** [comment]
@macfergus, I have bulid and run baduk successfully through mysy2,but I don't know how to use the library in windows,the import is no use.
![image](https://user-images.githubusercontent.com/43291475/83221568-bdda4f00-a1a8-11ea-92da-f1449ffb2a3c.png)
![image](https://user-images.githubusercontent.com/43291475/83221646-f0844780-a1a8-11ea-8cf9-b0facd5b01bf.png)
